### PR TITLE
JMeter implementation of registry pattern

### PR DESCRIPTION
**Pull request by diegomarangoni**
EES-4627
This PR implement new registry pattern into JMeter backend

### CODE REVIEW

**Comment by s4nji** [comment]
👍

File: pkg/backends/jmeter/backend.go [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	coreListersV1 "k8s.io/client-go/listers/core/v1"
+
+	"github.com/hellofresh/kangal/pkg/backends/internal"
+	"github.com/hellofresh/kangal/pkg/core/helper"
+	loadTestV1 "github.com/hellofresh/kangal/pkg/kubernetes/apis/loadtest/v1"
+	clientSetV "github.com/hellofresh/kangal/pkg/kubernetes/generated/clientset/versioned"
+
+	"go.uber.org/zap"
```
Comment (boekkooi-fresh) [inline pkg/backends/jmeter/backend.go:16]: Did goimport formatting go bad again?

File: pkg/backends/jmeter/backend.go [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+						zap.String("pod", pod.Name),
+						zap.String("namespace", namespace.GetName()),
+					)
+					loadTestStatus.Phase = loadTestV1.LoadTestErrored
+					return nil
+				}
+			}
+
+			return nil
+		}
```
Comment (boekkooi-fresh) [inline pkg/backends/jmeter/backend.go:266]: Can we extract this into a `func`?

File: pkg/backends/jmeter/backend.go [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+func (b *Backend) SyncStatus(ctx context.Context, loadTest loadTestV1.LoadTest, loadTestStatus *loadTestV1.LoadTestStatus) error {
+	// Get the Namespace resource
+	namespace, err := b.namespaceLister.Get(loadTestStatus.Namespace)
+	if err != nil {
+		// The LoadTest resource may no longer exist, in which case we stop
+		// processing.
+		if kerrors.IsNotFound(err) {
+			loadTestStatus.Phase = loadTestV1.LoadTestFinished
+			return nil
+		}
```
Comment (boekkooi-fresh) [inline pkg/backends/jmeter/backend.go:215]: Are we missing an err return?
```suggestion
		}
		return err
```

File: pkg/backends/jmeter/jmeter.go [hunk 4]
```diff
@@ -53,223 +45,54 @@ func New(
 	logger *zap.Logger,
 	namespacesLister coreListersV1.NamespaceLister,
 	reportURL string,
-	podAnnotations, namespaceAnnotations map[string]string,
+	podAnnotations, _ map[string]string,
```
Comment (boekkooi-fresh) [inline pkg/backends/jmeter/jmeter.go:48]: Why is `namespaceAnnotations` no longer used?

File: pkg/backends/jmeter/jmeter.go [hunk 5]
```diff
@@ -53,223 +45,54 @@ func New(
...
-	// the Job status
-	c.loadTest.Status.Phase = getLoadTestPhaseFromJob(job.Status)
-	c.loadTest.Status.JobStatus = job.Status
-
-	return nil
+	return c.backend.SyncStatus(ctx, *c.loadTest, &c.loadTest.Status)
 }
 
-func (c *JMeter) createPodsWithTestdata(ctx context.Context, configMaps []*coreV1.ConfigMap, loadtest *loadTestV1.LoadTest, namespace string) error {
+func (b *Backend) createPodsWithTestdata(ctx context.Context, configMaps []*coreV1.ConfigMap, loadTest *loadTestV1.LoadTest, namespace string) error {
```
Comment (boekkooi-fresh) [inline pkg/backends/jmeter/jmeter.go:81]: I think you forgot to move this to the right file.

File: pkg/backends/jmeter/resources.go [hunk 6]
```diff
@@ -56,7 +54,7 @@ var (
 )
 
 // NewJMeterSettingsConfigMap creates a new configmap which holds jmeter config
-func (c *JMeter) NewJMeterSettingsConfigMap() *coreV1.ConfigMap {
+func (b *Backend) NewJMeterSettingsConfigMap() *coreV1.ConfigMap {
```
Comment (boekkooi-fresh) [inline pkg/backends/jmeter/resources.go:57]: Why was this changed? This breaks BC at the moment right? 
(Same question for other `func (c *JMeter)` to `func (b *Backend) ` changes.)

File: pkg/backends/jmeter/jmeter.go (hunk 4 above)
Comment (diegomarangoni) [reply pkg/backends/jmeter/jmeter.go:48]: Namespace are created by controller itself, so backends don't need it.
This will be removed on the future

File: pkg/backends/jmeter/resources.go (hunk 6 above)
Comment (diegomarangoni) [reply pkg/backends/jmeter/resources.go:57]: `JMeter` now only have aliases to `Backend` functions, so this needs to be moved to new `Backend` struct as well.
Once the switch to this new registry pattern is done, all this `JMeter` struct can be deleted.

Comment (boekkooi-fresh) [reply pkg/backends/jmeter/resources.go:57]: I know but right now the tests no longer cover `JMeter` meaning people can break tests easily for `JMeter` but keep `Backend` tests passing. 
This in turn leads to other issues. I would advice to move the tests once you remove `JMeter` and we switch to using `Backend` directly

Comment (diegomarangoni) [reply pkg/backends/jmeter/resources.go:57]: That's the plan, but I also had to adapt the test, I will tag you on the related code.

File: pkg/backends/jmeter/resources_test.go [hunk 7]
```diff
@@ -93,19 +93,20 @@ func TestGetNamespaceFromInvalidName(t *testing.T) {
...
+				Tag:   defaultMasterImageTag,
+			},
+			WorkerConfig: loadtestv1.ImageDetails{
+				Image: defaultWorkerImageName,
+				Tag:   defaultWorkerImageTag,
 			},
 		},
+	}
+
+	c := &Backend{
```
Comment (diegomarangoni) [inline pkg/backends/jmeter/resources_test.go:109]: @boekkooi-fresh note that here I had to change from `JMeter` to `Backend` instead, to actually test those implementations.

File: pkg/backends/jmeter/backend.go [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+	if err != nil {
+		return err
+	}
+
+	if len(pods.Items) != int(*loadTest.Spec.DistributedPods) {
+		return nil
+	}
+
+	for _, pod := range pods.Items {
+		if pod.Status.Phase != "Running" {
```
Comment (aleksandrzhiliaev) [inline pkg/backends/jmeter/backend.go:240]: Should those pod status phases be consts?

File: pkg/backends/jmeter/backend.go [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+				loadTestStatus.Phase = loadTestV1.LoadTestFinished
+				return nil
+			}
+			loadTestStatus.Phase = getLoadTestStatusPhaseByPod(pod)
+			if loadTestV1.LoadTestErrored == loadTestStatus.Phase {
+				b.logger.Info(
+					"One of containers is unhealthy, marking LoadTest as errored",
+					zap.String("loadTest", loadTest.GetName()),
+					zap.String("pod", pod.Name),
+					zap.String("namespace", namespace.GetName()),
```
Comment (aleksandrzhiliaev) [inline pkg/backends/jmeter/backend.go:253]: Here `namespace` and load test name are identical.

File: pkg/backends/jmeter/jmeter.go [hunk 10]
```diff
@@ -53,243 +41,35 @@ func New(
...
-		}
-		c.logger.Info(
-			"Created JMeter resources",
-			zap.String("LoadTest", c.loadTest.GetName()),
-			zap.String("namespace", c.loadTest.Status.Namespace),
-		)
-
-	}
-	return nil
+	return c.backend.Sync(ctx, *c.loadTest, c.reportURL)
```
Comment (aleksandrzhiliaev) [inline pkg/backends/jmeter/jmeter.go:68]: I would also add debug logging on each `CheckOrCreateResources` or we can move `debug` log to `Sync` method.

File: pkg/backends/jmeter/jmeter.go [hunk 11]
```diff
@@ -53,243 +41,35 @@ func New(
...
-	if status.Active > 0 {
-		return loadTestV1.LoadTestRunning
-	}
-
-	if status.Succeeded == 0 && status.Failed == 0 {
-		return loadTestV1.LoadTestStarting
-	}
-
-	return loadTestV1.LoadTestFinished
+	return c.backend.SyncStatus(ctx, *c.loadTest, &c.loadTest.Status)
```
Comment (aleksandrzhiliaev) [inline pkg/backends/jmeter/jmeter.go:74]: I would also add debug logging on each `CheckOrUpdateStatus` or we can move debug log to `SyncStatus` method.

**Review by aleksandrzhiliaev — COMMENTED** [review]
Left small suggestions. Do we cover that JMeter backend by tests?

**Comment by diegomarangoni** [comment]
> Left small suggestions. Do we cover that JMeter backend by tests?

Yes, we do

File: pkg/backends/jmeter/jmeter.go (hunk 10 above)
Comment (diegomarangoni) [reply pkg/backends/jmeter/jmeter.go:68]: which log are you referring to? I didn’t remove any log, or did?

File: pkg/backends/jmeter/jmeter.go (hunk 11 above)
Comment (diegomarangoni) [reply pkg/backends/jmeter/jmeter.go:74]: which log are you referring to? I didn’t remove any log, or did?

File: pkg/backends/jmeter/jmeter.go (hunk 10 above)
Comment (aleksandrzhiliaev) [reply pkg/backends/jmeter/jmeter.go:68]: I think we didn't have any. I thought maybe as addition, might be useful when you debugging controller.

Comment (boekkooi-fresh) [reply pkg/backends/jmeter/jmeter.go:68]: Sounds like a good idea for another Ticket + PR

**Comment by aleksandrzhiliaev** [comment]
+1
